generator/templates/wire/WireClient.cpp

```diff
@@ -411,6 +510,49 @@ namespace wire {
                         return true;
                     }
                 {% endfor %}
+
+                bool HandleBufferMapReadAsyncCallback(const uint8_t** commands, size_t* size) {
+                    const auto* cmd = GetCommand<ReturnBufferMapReadAsyncCallbackCmd>(commands, size);
+                    if (cmd == nullptr) {
+                        return false;
+                    }
+
+                    auto* buffer = device->buffer.GetObject(cmd->bufferId);
+                    uint32_t bufferSerial = device->buffer.GetSerial(cmd->bufferId);
+
+                    //* The buffer might have been deleted or recreated so this isn't an error.
+                    if (buffer == nullptr || bufferSerial != cmd->bufferSerial) {
+                        return true;
+                    }
+
+                    //* The requests can have been deleted via an Unmap so this isn't an error.
+                    auto requestIt = buffer->readRequests.find(cmd->requestSerial);
+                    if (requestIt == buffer->readRequests.end()) {
+                        return true;
+                    }
+
+                    auto request = requestIt->second;
+
+                    //* On success, we copy the data locally because the IPC buffer isn't valid outside of this function
+                    if (cmd->status == NXT_BUFFER_MAP_READ_STATUS_SUCCESS) {
+
+                        //* The server didn't send the right amount of data, this is an error and could cause
+                        //* the application to crash if we did call the callback.
+                        if (request.size != cmd->dataLength) {
+                            return false;
+                        }
+
+                        if (buffer->mappedData != nullptr) {
+                            return false;
+                        }
+                        buffer->mappedData = malloc(request.size);
+                        memcpy(buffer->mappedData, cmd->GetData(), request.size);
+                    }
+
+                    request.callback(static_cast<nxtBufferMapReadStatus>(cmd->status), buffer->mappedData, request.userdata);
```

Comment:
Right, but wouldn't a failed request call the callback with an error status and the previous `buffer->mappedData` which would point to valid memory if `Unmap` has not been called?